### ISSUE TITLE
Documentation CI: 147 'missing documentation' warnings on pub items

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#26]
## Symptom

`cargo +nightly doc --features training --no-deps` now produces 0 `unresolved link` warnings (fixed in #19 / PR #24), but the Documentation CI job is still red because of ~147 `missing documentation` warnings on public items. CI uses `RUSTDOCFLAGS="-D warnings"` (`.github/workflows/ci.yml:115-120`), so every warning is fatal.

These warnings predate #19 and were already failing CI before any of the recent work.

## Decision needed

Two options:

1. **Add the docs** — write rustdoc for every public item flagged. Bulk task, low risk, high churn.
2. **Relax the lint** — change CI to not treat `missing-docs` as a hard error, or add `#![allow(missing_docs)]` at the crate root with the rationale documented.

The choice depends on whether public-item documentation is a project value or an aspirational gate that has been red for a while.

## Reproduction

```bash
cargo +nightly doc --features training --no-deps 2>&1 | grep -c 'missing documentation'
```

## Acceptance

Whichever path is chosen: the Documentation CI job goes green on `main`.

## Discovery

Flagged by the curator and judge for #19 ([judge comment](https://github.com/rjwalters/thrust/pull/24)). Explicitly out of scope for #19, which fixed the `unresolved link` class only.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating this issue per user request. Verified the count, mapped the warnings to source files, and sampled the offending items to inform the fork-in-the-road decision.

### Verification

Confirmed on `master` (commit `7ba44e3`) with libtorch from `./scripts/setup-libtorch.sh`:

| Command | Count |
|---------|-------|
| `cargo +nightly doc --features training --no-deps` | **143** |
| `cargo +nightly doc --no-deps --all-features` (matches CI in `.github/workflows/ci.yml:115-130`) | **147** |

The body's "147" is the CI-equivalent number. The 4-warning delta is `src/wasm.rs` items gated behind the `wasm` feature.

### Per-file concentration (top 20)

```
  40  src/policy/universal_inference.rs
  16  src/env/games/snake/types.rs
  14  src/policy/inference.rs
  13  src/multi_agent/joint.rs
  12  src/inference/snake.rs
  11  src/buffer/rollout/storage.rs
  10  src/multi_agent/population.rs
   7  src/policy/mlp.rs
   6  src/optimize/space.rs
   5  src/multi_agent/matchmaking.rs
   4  src/wasm.rs
   3  src/multi_agent/messages.rs
   3  src/optimize/{bayesian,pareto,scheduler}.rs  (1 each)
   1  src/multi_agent/environment.rs
   1  src/env/games/snake/snake.rs
   1  src/env/games/snake.rs
```

The top 9 files account for **130/147 (88%)** of the warnings. The work is highly concentrated.

### Sample characterization of offending items

I read across the worst-affected files to classify each item as *real public API* vs *internal-feeling pub that wants `pub(crate)`*. Results:

**Genuinely public API (consumed externally, documenting is the right fix):**

- `env/games/snake/types.rs` — `Direction`, `Position`, `Cell`, `GameState` and their variants/fields. These are re-exported from `crate::env::games::snake` (see `src/env/games/snake.rs:19`) and used by examples like `examples/games/snake/export_snake_model.rs`. **Trivially documentable** (Up/Down/Left/Right, x/y, etc.).
- `policy/universal_inference.rs` — `Activation`, `Layer::{Linear,Conv2d,Activation,Flatten,Reshape,Residual}`, `InputSpec`, `OutputSpec`, `ModelMetadata`. These are `Serialize`/`Deserialize` structs that define a public JSON schema for trained models. Field names are part of the wire format. **Must be documented.**
- `inference/snake.rs` — `SnakeCNNInference` weight/bias fields. Used in struct-literal form by `examples/games/snake/export_snake_model.rs:131` and `export_snake_init.rs:117`, plus deserialized in `src/wasm.rs:257`. **Must be documented** (existing sibling fields already have one-line `// [shape]` annotations that can be promoted to `///` docs).
- `policy/inference.rs` — Same pattern as `inference/snake.rs`. Same call.
- `multi_agent/joint.rs` — `JointTrainerConfig` fields (`gamma`, `gae_lambda`, `clip_range`, `vf_coef`, `ent_coef`, etc.). These are user-tunable PPO hyperparameters set in training scripts (`examples/games/bucket_brigade/train_p3.rs`). **Must be documented.**
- `policy/mlp.rs` — `MlpConfig` (`num_layers`, `hidden_dim`, `use_orthogonal_init`, `activation`) and `Activation::{ReLU, Tanh}`. User-facing config. **Document.**
- `optimize/space.rs` — `Parameter::Continuous { name, min, max, log_scale }`, `Parameter::Discrete { name, values }` named-field enum variants. The doc comment lives on the variant, not the fields. **Document fields.**
- `wasm.rs` — `WasmCartPole::new`, `WasmSnake::new`, `WasmSnake::get_height` — these are `#[wasm_bindgen]`-exposed JS API. **Must be documented** for browser consumers.
- `multi_agent/population.rs` — `Agent::{fitness, games_played, total_reward}`, `LearningMode::OffPolicy { buffer_size }`, `Hybrid { on_policy_ratio }`, `PopulationStats` fields. **Document.**
- `multi_agent/matchmaking.rs` — 4 × `pub fn new() -> Self` constructors and `FitnessBased { window_size }`. **Document trivially.**
- `multi_agent/messages.rs` — `ControlMessage::{SaveCheckpoint, LoadCheckpoint, SetLearningRate}` named fields. **Document trivially.**
- `multi_agent/environment.rs:84` — `pub truncated: Vec<bool>` on `MultiAgentResult`; the sibling fields (`observations`, `rewards`, `terminated`) are already documented — one was missed. **Trivially close the gap.**
- `env/games/snake.rs:35` — `pub type SnakeEnvSingle = SnakeEnv` legacy alias. **One line.**
- `env/games/snake/snake.rs:118` — `Food::position`. **One line.**
- `buffer/rollout/storage.rs` — 11 × `pub fn observations() -> &[...]` accessor methods on `RolloutBuffer`. These are public read-only getters. Could go either way — see "Edge cases" below.

**Internal-feeling `pub` that wants `pub(crate)` (visibility tightening is the right fix):**

- `src/optimize/bayesian.rs:6` — `pub struct BayesianOptimizer;` — empty placeholder struct, file body is `// Placeholder for future Bayesian optimization implementation`.
- `src/optimize/pareto.rs:3` — `pub struct ParetoFrontier;` — empty, `// TODO: Implement Pareto frontier tracking`.
- `src/optimize/scheduler.rs:3` — `pub struct TrialScheduler;` — empty, `// TODO: Implement parallel trial execution with rayon`.

These three are re-exported from `src/optimize/mod.rs:37-39` and appear in the module's rustdoc example as if usable, but the structs themselves have no methods, no fields, no functionality. They are aspirational API. **Tightening to `pub(crate)` is misleading either way** — the honest fix is to either delete them until they have an implementation, or document them as `unimplemented!()`-style markers ("Not yet implemented; tracked in #NN").

### Edge case: `buffer/rollout/storage.rs` accessors

The 11 warnings here are on field-accessor methods (`observations()`, `actions()`, etc.) of `RolloutBuffer`. The struct itself is public; the accessors exist because the fields are private. Two reasonable paths:

- **(a)** Add one-line `/// Get the buffer's observations`-style doc on each. ~30 LoC change, trivially correct.
- **(b)** Tighten the accessors to `pub(crate)` if `RolloutBuffer` is only consumed inside the `train` module — but the buffer is part of the documented `prelude` ambition (`src/lib.rs:55-58`) and users will eventually want to introspect rollouts. (a) is the safer pick.

### N/M ratio

- **Tighten visibility / remove**: ~3 items (the empty `optimize` stubs)
- **Document**: ~144 items, of which ~120 are one-liners (variants, x/y fields, shaped tensor fields with existing inline `// [shape]` comments that just need to become `///`)

This is **not** a "lots of these should be `pub(crate)`" situation — it's a "the lint level was set in `src/lib.rs:16` (`#![warn(missing_docs)]`) before the items existed, and the items are mostly legitimate public API that nobody got around to documenting." About 80% of the work is mechanical promotion of existing `//` end-of-line comments to `///` doc comments.

### Recommendation: Option 1 (add the docs)

**Reasoning**:

1. **Most items are real public API.** Examples, WASM consumers, and serde JSON consumers already depend on these names. Relaxing the lint would degrade the rustdoc index for downstream users.
2. **The work is mechanical and concentrated.** 9 files contain 88% of the warnings. Most items are enum variants, struct fields, or `new()` constructors — single-line docs suffice.
3. **The crate-level intent is already documented.** `src/lib.rs:16` says `#![warn(missing_docs)]`. CI's `RUSTDOCFLAGS=\"-D warnings\"` (`.github/workflows/ci.yml:130`) makes that a hard gate. Removing the gate now would silently allow regressions as the API grows.
4. **The cost is low.** Even at 30 seconds per item (most are 5-10 seconds), 147 items is ~75 minutes of focused work — well within a single Builder PR if scoped tightly.

**If the human picks Option 2 anyway**, the lightest-touch escape is `#![allow(missing_docs)]` at `src/lib.rs:16` with a rationale comment, *not* a CI change — keep the rustdoc gate, just disable this lint until docs are filled in. That preserves the unresolved-link gate fixed by #19/#24.

### Suggested phasing if Option 1

If 147 items is too large for a single Builder PR (likely — touches 19 files, ~150-300 LOC churn), suggest splitting into 2-3 sub-issues by area:

- **Phase A (~75 items)**: `policy/` and `inference/` (universal_inference 40 + inference 14 + snake 12 + mlp 7 + storage 11 ≈ 84 items, 5 files). One PR.
- **Phase B (~40 items)**: `multi_agent/` (joint 13 + population 10 + matchmaking 5 + messages 3 + environment 1 = 32 items) + `optimize/space.rs` (6) ≈ 38 items, 5 files.
- **Phase C (~25 items)**: `env/games/snake/` (16 + 1 + 1 = 18) + `wasm.rs` (4) + `optimize/{bayesian,pareto,scheduler}.rs` (3, decide here whether to delete/document/`pub(crate)`) ≈ 25 items, 7 files.

Each phase is independently mergeable and keeps CI red until all three land — or merge the last phase with `loom:urgent` if you want CI green sooner.

## Affected Files

If Option 1 (document):
- `src/policy/universal_inference.rs` — 40 items
- `src/env/games/snake/types.rs` — 16 items
- `src/policy/inference.rs` — 14 items
- `src/multi_agent/joint.rs` — 13 items
- `src/inference/snake.rs` — 12 items
- `src/buffer/rollout/storage.rs` — 11 items
- `src/multi_agent/population.rs` — 10 items
- `src/policy/mlp.rs` — 7 items
- `src/optimize/space.rs` — 6 items
- `src/multi_agent/matchmaking.rs` — 5 items
- `src/wasm.rs` — 4 items
- `src/multi_agent/messages.rs` — 3 items
- `src/multi_agent/environment.rs` — 1 item
- `src/env/games/snake/snake.rs` — 1 item
- `src/env/games/snake.rs` — 1 item
- `src/optimize/bayesian.rs`, `src/optimize/pareto.rs`, `src/optimize/scheduler.rs` — 1 each (or delete/`pub(crate)`)

If Option 2 (relax the lint):
- `src/lib.rs:16` — change `#![warn(missing_docs)]` to `#![allow(missing_docs)]` with a rationale comment referencing this issue
- *Do not* touch `.github/workflows/ci.yml`; the `RUSTDOCFLAGS=\"-D warnings\"` gate still catches `unresolved link` (the regression class #19 fixed) and any new warning categories rustdoc adds.

## Test Plan

- [ ] **Documentation CI passes**: `RUSTDOCFLAGS=\"-D warnings\" cargo +nightly doc --no-deps --all-features` exits 0. This is the exact CI invocation from `.github/workflows/ci.yml:130`.
- [ ] **No regression on \`unresolved link\`**: `cargo +nightly doc --no-deps --all-features 2>&1 | grep -c 'unresolved link'` returns 0 (#19/#24 baseline).
- [ ] **Builder/test job still green**: `cargo test --features training` continues to pass (docs-only changes should not affect tests, but verify the doctests in `///` comments compile if any are added).
- [ ] **No new clippy warnings**: `cargo +nightly clippy --all-targets --all-features -- -D warnings` (Option 1 only — Option 2 doesn't touch source).
- [ ] **For Option 2 specifically**: confirm a new undocumented `pub` item added in a follow-up PR does *not* cause CI to fail (sanity check that the relaxation actually relaxed).

## Notes for the human

- This issue should remain at `loom:curated` (not `loom:issue`) until you make the option call. If you pick Option 1 and want to phase it, add comments saying "approve Phase A" or split into sub-issues; if Option 2, the work is trivial enough that a single `loom:issue` will do.
- The `unresolved link` class was fixed in #19/PR #24, so this is the *only* outstanding rustdoc lint gate.
- Sample documentation effort: I sampled `src/env/games/snake/types.rs` (16 warnings); a complete fix is ~25 lines of one-line `///` comments. Extrapolating, the full Option-1 patch is ~300-400 LOC of pure doc-comment additions.
